httpbp/server_test.go

```diff
@@ -427,3 +433,283 @@ func TestPanicRecovery(t *testing.T) {
+							t.Fatalf("missing header %q", wantKey)
+						} else if diff := cmp.Diff(v, wantValue, cmpopts.SortSlices(func(a, b string) bool {
+							return a < b
+						})); diff != "" {
+							t.Fatalf("header %q values mismatch (-want +got):\n%s", wantKey, diff)
+						}
+					}
+					return nil
+				},
+			},
+		},
+		Middlewares: []httpbp.Middleware{
+			httpbp.ServerHeaderBPMiddleware("originHTTPBPV0"),
+		},
+	})
+	if err != nil {
+		t.Fatalf("failed to create test downstreamServer: %v", err)
+	}
+	t.Cleanup(func() {
+		downstreamServer.Close()
+	})
+	go downstreamServer.Serve()
+
+	downstreamBaseURL, err := url.Parse("http://" + downstreamServer.Baseplate().GetConfig().Addr + "/")
+	if err != nil {
+		t.Fatalf("failed to parse test originServer base URL: %v", err)
+	}
+
+	downstreamClient, err := httpbp.NewClient(
+		httpbp.ClientConfig{
+			Slug:         "downstreamHTTPBPV0",
+			InternalOnly: true,
+		},
+		withBaseURL(downstreamBaseURL),
+	)
+	if err != nil {
+		t.Fatalf("failed to create test client: %v", err)
+	}
+
+	originServer, err := httpbp.NewBaseplateServer(httpbp.ServerArgs{
+		Baseplate: bp,
+		Endpoints: map[httpbp.Pattern]httpbp.Endpoint{
+			"/say-hello": {
+				Name:    "say-hello",
+				Methods: []string{http.MethodGet},
+				Handle: func(ctx context.Context, writer http.ResponseWriter, request *http.Request) error {
+					for wantKey, wantValue := range expectedHeaders {
+						if v := request.Header.Values(wantKey); len(v) == 0 {
+							t.Fatalf("missing header %q", wantKey)
+						} else if diff := cmp.Diff(v, wantValue, cmpopts.SortSlices(func(a, b string) bool {
+							return a < b
+						})); diff != "" {
+							t.Fatalf("header %q values mismatch (-want +got):\n%s", wantKey, diff)
+						}
+					}
+
+					req, err := http.NewRequest(
+						http.MethodGet,
+						downstreamBaseURL.JoinPath("say-hello").String(),
+						nil,
+					)
+					if err != nil {
+						t.Fatalf("creating request: %v", err)
+					}
+
+					resp, err := downstreamClient.Do(req)
+					if err != nil {
+						t.Fatalf("unexpected error: %v", err)
+					}
+					if resp.StatusCode != http.StatusOK {
+						t.Fatalf("unexpected status code: %d", resp.StatusCode)
+					}
+
+					invalidReq, err := http.NewRequest(
+						http.MethodGet,
+						downstreamBaseURL.JoinPath("say-hello").String(),
+						nil,
+					)
+					if err != nil {
+						t.Fatalf("creating request: %v", err)
+					}
+					invalidReq.Header.Set("x-bp-test", "bar")
+
+					if _, err := downstreamClient.Do(req); !errors.Is(err, headerbp.ErrNewInternalHeaderNotAllowed) {
+						t.Fatalf("error mismatch, want %v, got %v", headerbp.ErrNewInternalHeaderNotAllowed, err)
+					}
+					return nil
+				},
+			},
+		},
+		Middlewares: []httpbp.Middleware{
+			httpbp.ServerHeaderBPMiddleware("originHTTPBPV0"),
+		},
+	})
+	if err != nil {
+		t.Fatalf("failed to create test originServer: %v", err)
+	}
+	t.Cleanup(func() {
+		originServer.Close()
+	})
+	go originServer.Serve()
```

Comment:
you can probably add a small sleep after this `go` to give it time to actually start the server, not sure if it's worth it though 🤷 looks like the failed test is just very unlucky.